### DISCUSSION TITLE
Imperative Representation of Terms

### DISCUSSION THREAD

**Discussion by Bohun9** [ppolesiuk/type-systems-notes#78] [Q&A] [answered]
I have some doubts about it:
1. Why there is no string identifier in definition of uvar? There can be many unification variables and we have to distinguish between them.
2. I guess implementation should use some module with abstract data types.
3. What are the benefits of using tp_view? I think the main benefit of using mutation is to replace type environment and substitution in the original algorithm with a global type environment, right?

**Comment by ppolesiuk** [comment, marked as answer]
1. It is not necessary. In this basic algorithm we only need checking equality between unification variables, and it is possible to check equality between mutable cells using physical equality (`==` in OCaml). In more practical setup it would be useful to have some order on unification variable, e.g. for constructing finite sets. To achieve this, we can represent unification variables as pairs of mutable cell and some element of some ordered type. But string is an overkill: most of unification variables are freshly generated by an algorithm and it is hard to generate fresh strings. Moreover, user may use the same name for two different unification variables. Natural numbers freshly generated by an ordinary counter is a better choice. If we want to identify unification variables by some names provided by a programmer, we can have some additional data structure (as a part of an environment) that links names with unification variables.
2. Using modules with abstract data types is a good design choice, as it enforce the user to always use `view` that may change already substituted unification variables to their values. However, abstract data types are not available in all languages, and the same thing may be enforced by other techniques (e.g., static analysis, code review, a hope that everybody reads documentation, etc.).
3. Type `tp_view` together with `view` function reveals the top most construct of the term, while type `tp` should be abstract. We use mutable state to efficiently implement global substitution. Mutable implementation of global type environment is an orthogonal issue, that we didn't covered on this lecture.
